solver/llbsolver/solver.go

```diff
@@ -308,6 +315,7 @@ func inlineCache(ctx context.Context, e remotecache.Exporter, res solver.CachedR
 			digests = append(digests, desc.Digest)
 		}
 
+		ctx = withDescHandlerCacheOpts(ctx, workerRef.ImmutableRef)
```

Comment:
I don't understand this. Why would cache export need access of `descHandlers` from `workerRef`? If it needs to unlazy anything that it is exporting cache for then it should be associated with the record that is being exported. There can be many more refs that are exported in cache, not just the final result ref.

Reply:
Are you saying that to call `immutableRef.GetRemote()` you need to pass deschandlers in context and it will error otherwise? Why is that? We are not reading ref from disk. We already have an instance and it is already initialized with deschandlers.

Reply:
> Are you saying that to call immutableRef.GetRemote() you need to pass deschandlers in context and it will error otherwise?


Yes, for example:
1. [This line calls ExportTo](https://github.com/moby/buildkit/blob/f421e242918320dbeeac982752de67e79fda9a49/solver/llbsolver/solver.go#L249). It has access to the ref with all the cache opts loaded up, but that ref is not passed to `ExportTo`.
2. [ExportTo calls LoadRemotes](https://github.com/moby/buildkit/blob/f421e242918320dbeeac982752de67e79fda9a49/solver/exporter.go#L114)
3. [LoadRemotes loads a new instance of the ref being exported using the result ID.](https://github.com/moby/buildkit/blob/f421e242918320dbeeac982752de67e79fda9a49/worker/cacheresult.go#L70-L77) This is where the ref gets re-read from disk and thus needs to have access to desc handler cache opts or else it gets an error.

If there's a simple way to safely pass the already loaded ref at step `1.` above to the later calls so that they don't need to reload the ref from disk, that seems like a better solution. However, I didn't want to embark on refactoring this tangentially related code immediately, hence the current solution. If you have specific suggestions on how to approach this a different way I can try them out.